### PR TITLE
WIP: 8051: handle any mov case

### PR DESCRIPTION
**Pull request by gogo2464**
**Checklist**

- [ ] Closing issues: #issue
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**

This is a great refactoring and a great improvement. Not ready yet. I need to check any possible case. For example, it does not work for 'mov 0x0f, a' yet.

### CODE REVIEW

**Comment by trufae** [comment]
Screenshot 2021-12-30 at 21 58 59

**Comment by gogo2464** [comment]
@trufae does it respect and improve C iso ?

**Comment by trufae** [comment]
Whats "improve the c iso"?

**Comment by gogo2464** [comment]
> Whats "improve the c iso"?

I mean is my refactoring when I refatcor more readable than the code before?

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:446]: can you rename the function name, because it have nothing to do with what the comment describes. thiçs way the comment can be removed

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:564]: please open { at right not below

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:1004]: dont else after return. actually in this case you could just return true once at the end. but the whole code looks very spaguetti and can be probably simplfied a lot theres a lot of copypaste

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by trufae** [comment]
> > Whats "improve the c iso"?
> 
> I mean is my refactoring when I refatcor more readable than the code before?

just say cleanup or refactor :P the C iso standard have nothing to do. but the coding style or developer guidelines do

File: libr/asm/arch/8051/8051_ass.c
Comment (gogo2464) [reply libr/asm/arch/8051/8051_ass.c:1004]: Just for return true or can I esle after a false return?

File: libr/asm/arch/8051/8051_ass.c
Comment (gogo2464) [reply libr/asm/arch/8051/8051_ass.c:446]: @trufae Do you think I modified this function too much? I did not want to change it's behavior.

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:343]:
```suggestion
static bool is_direct(char const*str)
```
```suggestion
static bool is_direct(char const *str) {
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:344]:
```suggestion
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [reply libr/asm/arch/8051/8051_ass.c:452]: If it needs to bw done it must be done its never too much (:

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:883]: Wrong indent

**Review by trufae — CHANGES_REQUESTED** [review]
Please add more tests to ensure future refactorings dont break it

**Comment by trufae** [comment]
This test is failing

`[XX] db/asm/8051 <asm> mov 0x55, 0x44`

**Comment by gogo2464** [comment]
I disassembled and reassembled a 8051 firmware of 0x7c4a bytes. Then I put radiff2 and diff and there was no difference. it is cool. It means there is no bug anymore!

I am going to add more tests for 8051 on another PR. See https://github.com/radareorg/radare2/issues/19481

File: libr/asm/arch/8051/8051_ass.c
Comment (Lazula) [inline libr/asm/arch/8051/8051_ass.c:343]:
```suggestion
static bool is_direct(char const *str) {
```

File: libr/asm/arch/8051/8051_ass.c
Comment (Lazula) [inline libr/asm/arch/8051/8051_ass.c:1041]: these checks seem to duplicate `is_reg()`/`register_number()`. it looks like there ought to be a function that returns 0 for [r0], 1 for [r1], and 2-9 for r0-r7. maybe `reg_offset()`?

File: libr/asm/arch/8051/8051_ass.c
Comment (Lazula) [inline libr/asm/arch/8051/8051_ass.c:353]:
```suggestion
r_return_val_if_fail (str, false);
return tolower (str[0]) == 'r'
	&& R_BETWEEN ('0', str[1], '7');
	&& !str[2];
```

File: libr/asm/arch/8051/8051_ass.c
Comment (Lazula) [inline libr/asm/arch/8051/8051_ass.c:977]: duplicate `is_reg()`?

File: libr/asm/arch/8051/8051_ass.c
Comment (Lazula) [inline libr/asm/arch/8051/8051_ass.c:1008]: could probably reduce this section by making the outside checks first and storing values to use in the inner checks

Comment (gogo2464) [reply libr/asm/arch/8051/8051_ass.c:1008]: I may have solved this. Could you get a look @Lazula please?

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:1249]:
```suggestion
static bool mnem_xrl(const char *arg, ut16 pc, ut8 *out, int out_len) {
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:1256]: out[0] = 0x63; // if you change the function signature to ut8 *out instead of **out

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:1422]: why carg is 2,1,0 ? be consistent. also those lines break the coding style as each statement should be in a separate line.

you can use `R_FREE (arg[1]);` instead of (free(arg[1]);arg[1]=0;) . also the `carg is just a local array, it doesnt needs to be deinitialized. and as long as its returning right away and all those are local vars you can just call `free` and the rest is unnecesary

File: libr/asm/arch/8051/8051_disas.c
Comment (trufae) [inline libr/asm/arch/8051/8051_disas.c:26]:
```suggestion
		char* disasm = NULL;
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:1069]:
```suggestion
static bool mnem_movc(const char *arg, ut16 pc, ut8 *out, int outlen) {
```

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by gogo2464** [comment]
I agree with you @trufae but I need to change the signature of any mnem_something then. Else I will get an error.

**Comment by Lazula** [comment]
He means that the mnem_ functions are required to have the signature `bool (*)(const char**, uint16_t, uint8t **)`, so changing it results in an error.

**Comment by trufae** [comment]
I think im fine to merge this but not for today's release. I can fix those function signatrures after merging. 

we have tons of prs waiting for this release. So fun time is coming back

**Comment by gogo2464** [comment]
Seems ready to merge then.

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:552]: don't else after return
```suggestion
	}
	if (!r_str_casecmp (reg, "@r1") || !r_str_casecmp (reg, "[r1]")) {
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:554]: dont else after return

File: libr/asm/arch/8051/8051_disas.c
Comment (trufae) [inline libr/asm/arch/8051/8051_disas.c:26]: i think those register aliases shouldnt be removed. or am i missing something :?

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/asm/arch/8051/8051_disas.c
Comment (gogo2464) [reply libr/asm/arch/8051/8051_disas.c:26]: This part of the code is only used by replace register. And I removed replace register because it does not really points to BANK and because it break the reassembly

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:557]:
```suggestion
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:556]:
```suggestion
	return register_number (reg) + 2;
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:1260]:
```suggestion
		}
		if (offset < 10) {
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:353]:
```suggestion
	return tolower (str[0]) == 'r' && R_BETWEEN ('0', str[1], '7') && !str[2];
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:354]:
```suggestion
```

File: libr/asm/arch/8051/8051_ass.c
Comment (trufae) [inline libr/asm/arch/8051/8051_ass.c:355]:
```suggestion
```

**Review by trufae — APPROVED** [review]

**Pull request merged**
